File: examples/typescript/src/bring-your-own-agent.ts
import { Marona, type MCPTool } from "marona";

import { requiredApiKey, selectedApp } from "./config.js";

export interface CapabilityConnection {
  listTools(): MCPTool[];
  callTool(
    name: string,
    input: Record<string, unknown>,
    options?: { userId?: string; conversationId?: string },
  ): Promise<Record<string, unknown>>;
}

export class ExistingAgentAdapter {
  constructor(private readonly connection: CapabilityConnection) {}

  availableTools(): MCPTool[] {
    return this.connection.listTools();
  }

  executeSelectedTool(
    name: string,
    input: Record<string, unknown>,
    identity: { userId: string; sessionId: string },
  ): Promise<Record<string, unknown>> {
    return this.connection.callTool(name, input, {
      userId: identity.userId,
      conversationId: identity.sessionId,
    });
  }
}

export async function runBringYourOwnAgent(): Promise<void> {
  const marona = new Marona({ apiKey: requiredApiKey() });
  const app = selectedApp();
  const connection = await marona.hub.connect(app ? { apps: [app] } : {});
  const externalAgent = new ExistingAgentAdapter(connection);

  console.log("The external framework can map these neutral tools:");
  for (const tool of externalAgent.availableTools()) {
    console.log(`- ${tool.name}`);
  }
  console.log("Tool selection stays external; capability execution remains governed by Marona.");
}

File: examples/python/src/marona_agents_examples/bring_your_own_agent.py
"""Keep an existing orchestrator while Marona governs capabilities."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Protocol

from marona import Marona

from .config import required_api_key, selected_app


class CapabilityConnection(Protocol):
    """Small boundary an existing framework needs from Marona."""

    def list_tools(self) -> list[dict[str, Any]]: ...

    def call_tool(
        self,
        name: str,
        arguments: dict[str, Any],
        *,
        user_id: str | None = None,
        conversation_id: str | None = None,
    ) -> Any: ...


@dataclass(slots=True)
class ExistingAgentAdapter:
    """Example adapter owned by an external agent framework."""

    connection: CapabilityConnection

    def available_tools(self) -> list[dict[str, Any]]:
        return self.connection.list_tools()

    def execute_selected_tool(
        self,
        name: str,
        arguments: dict[str, Any],
        *,
        user_id: str,
        session_id: str,
    ) -> Any:
        return self.connection.call_tool(
            name,
            arguments,
            user_id=user_id,
            conversation_id=session_id,
        )


def main() -> None:
    marona = Marona(api_key=required_api_key())
    try:
        app = selected_app()
        connection = marona.hub.connect(apps=[app] if app else None)
        external_agent = ExistingAgentAdapter(connection)

        print("The external framework can map these neutral tools:")
        for tool in external_agent.available_tools():
            print(f"- {tool.get('name', 'unnamed_tool')}")
        print(
            "Tool selection remains in the external framework; "
            "execution remains governed by Marona."
        )
    finally:
        marona.close()


if __name__ == "__main__":
    main()
